### ISSUE TITLE
What would it take to implement transparency for custom controls?

### ISSUE THREAD

**Issue by lszl84** [wxWidgets/wxWidgets#23412]
### Description

Suppose I want to write a library with custom rounded buttons for wxWidgets. Here's a simple idea:

```cpp
#include <wx/wx.h>
#include <wx/dcbuffer.h>

class MyApp : public wxApp
{
public:
    virtual bool OnInit();
};

wxIMPLEMENT_APP(MyApp);

class MyFrame : public wxFrame
{
public:
    MyFrame(const wxString &title, const wxPoint &pos, const wxSize &size);
};

bool MyApp::OnInit()
{
    MyFrame *frame = new MyFrame("Hello World", wxDefaultPosition, wxDefaultSize);
    frame->Show(true);
    return true;
}

class MyButton : public wxWindow
{
public:
    MyButton(wxWindow *parent, wxWindowID id, const wxPoint &pos, const wxSize &size, long style = 0, const wxString &name = wxPanelNameStr)
        : wxWindow(parent, id, pos, size, style, name)
    {
        this->Bind(wxEVT_PAINT, &MyButton::OnPaint, this);
    }

    void OnPaint(wxPaintEvent &event)
    {
        wxPaintDC dc(this);

        wxColor color = wxColor(100, 100, 200);

        dc.SetPen(wxPen(color, FromDIP(10)));
        dc.SetBrush(wxBrush(color));

        auto rect = this->GetClientRect();
        rect.Deflate(FromDIP(10));
        dc.DrawRoundedRectangle(rect, FromDIP(20));
    }
};

MyFrame::MyFrame(const wxString &title, const wxPoint &pos, const wxSize &size)
    : wxFrame(nullptr, wxID_ANY, title, pos, size)
{

    auto sizer = new wxBoxSizer(wxVERTICAL);

    auto panel = new wxPanel(this);
    panel->SetBackgroundColour({200, 100, 100});

    auto custom = new MyButton(panel, wxID_ANY, wxDefaultPosition, wxDefaultSize);

    auto panelSizer = new wxBoxSizer(wxVERTICAL);
    panelSizer->Add(custom, 1, wxEXPAND | wxALL, FromDIP(50));
    panel->SetSizer(panelSizer);

    sizer->Add(panel, 1, wxEXPAND);
    this->SetSizer(sizer);
}
```

Works well on GTK or Mac but not on Windows. Experimenting with clearing the background with a transparent brush yields mixed results (weird paint artifacts).

Setting the background to the parent's color might be a workaround, but what if the user wants to place my custom button on top of their custom bitmap control? As the library author, I don't have access to the user's controls, so I cannot render its part in my `wxPaintDC.`

TL;DR, it seems such a basic thing as creating a custom rounded rect button is not possible on wxWidgets as of now. What would it take to implement transparency and compositing on Windows, so it can be done?

<img width="1326" alt="Screenshot 2023-04-01 at 13 25 23" src="https://user-images.githubusercontent.com/42072702/229286334-948df5ec-a9a7-4324-9192-d5b997d341ec.png">
<img width="1712" alt="Screenshot 2023-04-01 at 13 24 34" src="https://user-images.githubusercontent.com/42072702/229286338-92fd7a6e-6134-45d7-a5d4-09dc731c195b.png">
<img width="1712" alt="Screenshot 2023-04-01 at 13 24 09" src="https://user-images.githubusercontent.com/42072702/229286343-a3733a4e-87d8-4688-94e7-2c1e6c3f0f6d.png">

**Comment by vadz** [comment]
You need to use `wxGraphicsContext` to have any chance of transparency working correctly.

**Comment by lszl84** [comment]
@vadz can you give an example? This also does not seem to work:

```cpp
//...

void OnPaint(wxPaintEvent &event)
    {
        wxPaintDC dc(this);

        std::unique_ptr<wxGraphicsContext> gc{wxGraphicsContext::Create(dc)};

        if (gc)
        {
            wxColor color = wxColor(100, 100, 200);

            gc->SetPen(wxPen(color, FromDIP(10)));
            gc->SetBrush(wxBrush(color));

            auto rect = this->GetClientRect();
            rect.Deflate(FromDIP(10));
            gc->DrawRoundedRectangle(rect.GetLeft(), rect.GetTop(), rect.GetWidth(), rect.GetHeight(), FromDIP(20));
        }
    }

//...
```

<img width="1712" alt="Screenshot 2023-04-01 at 20 14 06" src="https://user-images.githubusercontent.com/42072702/229307711-9376a079-5da0-43e1-a370-50521f73e1de.png">

**Comment by vadz** [comment]
Sorry, I'm not sure if it works even with wxGC, I was just saying that it certainly can't work when using GDI.

Are you testing this with 3.3 (which uses compositing under MSW too)? Also, does using alpha component for the colour/brush change anything?

**Comment by lszl84** [comment]
I've just tried 3.3, and the results are the same, unfortunately, as in 3.2.2.

Setting the alpha component for the color makes it appear lighter, but still the color of the control underneath does not come trough.

<img width="1712" alt="Screenshot 2023-04-01 at 20 56 34" src="https://user-images.githubusercontent.com/42072702/229309272-c4a9e9c2-7441-41db-a9b1-3b37a69f2218.png">

**Comment by MaartenBent** [comment]
Hi you can achieve this using `SetBackgroundStyle(wxBG_STYLE_TRANSPARENT);` and `WS_EX_TRANSPARENT`.
A quick test, not portable:

```c++
    MyButton(wxWindow* parent, wxWindowID id, const wxPoint& pos, const wxSize& size, long style = 0, const wxString& name = wxPanelNameStr)
        : wxWindow()
    {
        SetBackgroundStyle(wxBG_STYLE_TRANSPARENT);
        wxWindow::Create(parent, id, pos, size, style, name);

        int extendedStyle = GetWindowLong(GetHWND(), GWL_EXSTYLE);
        SetWindowLong(GetHWND(), GWL_EXSTYLE, extendedStyle | WS_EX_TRANSPARENT);

        this->Bind(wxEVT_PAINT, &MyButton::OnPaint, this);
    }
```

**Comment by lszl84** [comment]
@MaartenBent Thank you, this works beautifully!

Maybe it would be a good idea to execute these lines in the wxWidgets code if the user sets the background style to transparent on Windows so that the behavior is consistent across platforms.

```cpp
int extendedStyle = GetWindowLong(GetHWND(), GWL_EXSTYLE);
SetWindowLong(GetHWND(), GWL_EXSTYLE, extendedStyle | WS_EX_TRANSPARENT);
```

What do you think? Judging by the comments on my tutorials, this issue seems to be quite frustrating to people ;)

**Comment by vadz** [comment]
We probably need to map `wxBG_STYLE_TRANSPARENT` to `WS_EX_TRANSPARENT`. But note that this will only work in 3.3 with `WS_EX_COMPOSITED`, not in 3.2.

**Comment by lszl84** [comment]
Actually, Maarten's solution works for me in 3.2:

<img width="1712" alt="Screenshot 2023-04-02 at 11 33 39" src="https://user-images.githubusercontent.com/42072702/229344765-21fb5637-ef9b-4c75-8f1f-d455d23f6cc7.png">

**Comment by MaartenBent** [comment]
Indeed, if I disable `WS_EX_COMPOSITED` it still works. I can add something like this, so `WS_EX_TRANSPARENT` is added when `wxBG_STYLE_TRANSPARENT` is used:

```diff
diff --git "a/src/msw/window.cpp" "b/src/msw/window.cpp"
index 4520968686..5f4aa7351d 100644
--- "a/src/msw/window.cpp"
+++ "b/src/msw/window.cpp"
@@ -1641,6 +1641,9 @@ WXDWORD wxWindowMSW::MSWGetStyle(long flags, WXDWORD *exstyle) const
             *exstyle |= WS_EX_CONTROLPARENT;
         }
 #endif // __WXUNIVERSAL__
+
+        if ( GetBackgroundStyle() == wxBG_STYLE_TRANSPARENT )
+            *exstyle |= WS_EX_TRANSPARENT;
     }
 
     return style;
```

**Comment by vadz** [comment]
I admit I'm quite surprised that it works without compositing. Does it also refresh correctly when resizing the window?

Be it as may, it would certainly be nice to support `wxBG_STYLE_TRANSPARENT` in master. We just need to also override `IsTransparentBackgroundSupported()` in `wxWindowMSW` and also update the docs to mention that it's supported there too.

**Comment by lszl84** [comment]
@vadz It does refresh correctly when resizing with `wxFULL_REPAINT_ON_RESIZE`
